src/Input/PathParam.php

```diff
@@ -0,0 +1,126 @@
+ */
+
+declare(strict_types=1);
+
+namespace Laminas\Cli\Input;
+
+use InvalidArgumentException;
+use RuntimeException;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Question\Question;
+
+use function array_map;
+use function file_exists;
+use function gettype;
+use function in_array;
+use function is_dir;
+use function is_string;
+use function preg_replace;
+use function scandir;
+use function sprintf;
+
+final class PathParam implements InputParamInterface
+{
+    use StandardQuestionTrait;
+
+    public const TYPE_DIR  = 'dir';
+    public const TYPE_FILE = 'file';
+
+    /**
+     * Whether or not the path provided must exist.
+     *
+     * @var bool
+     */
+    private $mustExist = false;
+
+    /**
+     * One of the TYPE_* constants
+     *
+     * @var string
+     */
+    private $type = self::TYPE_FILE;
+
+    public function __construct(string $name)
+    {
+        $this->name = $name;
+    }
+
+    public function getOptionMode(): ?int
+    {
+        return InputOption::VALUE_REQUIRED;
+    }
+
+    public function getQuestion(): Question
+    {
+        $question = $this->createQuestion();
+
+        $question->setAutocompleterCallback(static function (string $userInput): array {
+            // Strip any characters from the last slash to the end of the string
+            // to keep only the last directory and generate suggestions for it
+            $inputPath = preg_replace('%(/|^)[^/]*$%', '$1', $userInput);
+            $inputPath = $inputPath === '' ? '.' : $inputPath;
+
+            // CAUTION - this example code allows unrestricted access to the
+            // entire filesystem. In real applications, restrict the directories
+            // where files and dirs can be found
+            $foundFilesAndDirs = is_dir($inputPath) ? scandir($inputPath) : [];
+
+            return array_map(static function (string $dirOrFile) use ($inputPath): string {
+                return $inputPath . $dirOrFile;
+            }, $foundFilesAndDirs);
+        });
+
+        $question->setValidator(function ($value) {
+            if ($value === null && ! $this->required) {
+                return null;
+            }
+
+            if (! is_string($value)) {
+                throw new RuntimeException(sprintf('Invalid value: string expected, %s given', gettype($value)));
+            }
+
+            if (! $this->mustExist) {
+                // No further checks needed
+                return $value;
+            }
+
+            if (! file_exists($value)) {
+                throw new RuntimeException('Path does not exist');
+            }
+
+            if ($this->type === self::TYPE_DIR && ! is_dir($value)) {
+                throw new RuntimeException('Path is not a valid directory');
+            }
+
+            return $value;
+        });
+
+        return $question;
+    }
+
+    public function setPathType(string $type): self
```

Comment:
If you read what he says, he also says they are fine _for builders_. In this case, we are doing a builder pattern in order to create a parameter. The reasons I chose this pattern here:

- It simplifies constructors, where we will generally have a LOT of optional arguments, and won't necessarily want to provide a bunch of `null` values or require knowledge of default values in order to set something far down the list. See the `Command::addOption()` and `Command::addArgument()` methods in symfony/console for what I was trying to avoid.

- It gives users better IDE support. Argument type hints are a PITA when you have a ton of them. However, IDEs are usually very succinct when offering up lists of methods, and when there's only one argument, the tooling is onobtrusive.

- It simplifies our interface for adding parameters to a command, as we can typehint on an interface, and there's exactly one permutation for adding.

Since parameters are basically options on steroids, there's a lot of pieces that can be used to define one:

- The name (required)
- A description (optional)
- A shortcut (optional)
- Whether or not it is required (optional, with default value)
- A default value (optional; defaults to `null`)

Different parameter types add more to the mix: choices need an array of valid choices; strings can optionally allow you to specify a pattern for validation; etc. Factory methods don't really cut it, as we need to allow specifying _all_ of these, and we then trade-off the builder pattern for an unwieldy list of arguments.

So, while I, too, am generally against fluent interfaces, this is exactly the sort of scenario they likely _should_ be used.